test/Unit/DependencyManagement/DependencyManagementTests.cs

```diff
@@ -177,5 +186,131 @@ public void TestManagedDependencyNoRequirementsFileShouldThrow()
+                var managedDependenciesFolderPath = GetManagedDependenciesPath(functionAppRoot);
+
+                var functionLoadRequest = GetFuncLoadRequest(functionFolderPath, true);
+
+                // Create DependencyManager and process the requirements.psd1 file at the function app root.
+                var dependencyManager = new TestDependencyManager();
+                dependencyManager.Initialize(functionLoadRequest);
+
+                // Set the dependencyManager to throw in the RunSaveModuleCommand call.
+                dependencyManager.ShouldThrow = true;
+
+                // Validate retry logic.
+                dependencyManager.InstallFunctionAppDependencies(null, _testLogger);
+
+                // Here we will get four logs: one that says that we are installing the
+                // dependencies, and three for failing to install the module.
+                bool correctLogCount = (_testLogger.FullLog.Count == 4);
+                Assert.True(correctLogCount);
+
+                // The first log should say "Installing FunctionApp dependent modules."
+                Assert.Contains(PowerShellWorkerStrings.InstallingFunctionAppDependentModules, _testLogger.FullLog[0]);
+
+                // The subsequent logs should contain the following:
+                for (int index = 1; index < _testLogger.FullLog.Count; index++)
+                {
+                    Assert.Contains("Fail to install module", _testLogger.FullLog[index]);
+                }
+
+                // Lastly, DependencyError should get set after unsuccessfully  retyring 3 times.
+                Assert.NotNull(dependencyManager.DependencyError);
+                Assert.Contains("Fail to install FunctionApp dependencies. Error:", dependencyManager.DependencyError.Message);
+            }
+            finally
+            {
+                TestCaseCleanup();
+            }
+        }
+    }
+
+    internal class TestDependencyManager : DependencyManager
+    {
+        public bool ShouldThrow { get; set; }
+
+        public string TestFileName { get; set; } = "ModuleInstalled.txt";
+
+        internal TestDependencyManager()
+        {
+        }
+
+        protected override void RunSaveModuleCommand(PowerShell pwsh, string repository, string moduleName, string version, string path)
+        {
+            if (ShouldThrow)
+            {
+                var errorMsg = $"Fail to install module '{moduleName}' version '{version}'";
+                throw new InvalidOperationException(errorMsg);
+            }
+
+            // Write a text file to the given path with the information of the module that was downloaded.
+            var message = string.Format(PowerShellWorkerStrings.ModuleHasBeenInstalled, moduleName, version);
+            var filePath = Path.Join(path, TestFileName);
+            File.WriteAllText(filePath, message);
+        }
```

Comment:
Do you really need a file? Why not just keep a list of messages in memory?

Reply:
We can keep the module name and version in memory. Test case updated.